### PR TITLE
Create a Tutorial base on deployment example

### PR DESCRIPTION
**Pull request by wtam2018**
Create a step by step guide/tutorial to setup multiple environment Gitops deployment example.

Fixes https://issues.redhat.com/browse/GITOPS-33

### CODE REVIEW

File: Example.md [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+* Set the following environment varialbes
+    * USER=\<OS user \>  (It should be automatically set by OS.)
+    * QUAY_USER=\<Ouay.io user\>
+    * GITHUB_USER=\<GitHub user\>
+    * DEPLOYMENT_PATH=deploy
+
+
+## Create Demo Git Repositories
+
+Login to Your GitHub account.  Create your demo Git repositories `taxi` and `taxi-stage-config`.  Then, import from the following repos, respectively.
```
Comment (bigkevmcd) [inline Example.md]: Might want to explain how to import these?

File: Example.md [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+
+Login to Your GitHub account.  Create your demo Git repositories `taxi` and `taxi-stage-config`.  Then, import from the following repos, respectively.   
+   * https://github.com/bigkevmcd/taxi
+   * https://github.com/bigkevmcd/taxi-stage-config
+
+These are your test repositories where you will submit Pull Requests, merge them to master in order to trigger Pipeline Runs.
+
+ ## Download your Gihub Access Token
+
+ The [`create-github-status-task`](#create-github-status-task) needs Github user's access token to update Github repo status.   You will create a Kubernates secret from the downloaded access token later in the Tutorial.
```
Comment (bigkevmcd) [inline Example.md]: s/Kubernates/Kubernetes/

File: Example.md [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+
+Create the following projects.
+
+```shell
+oc new-project ${USER}-dev-environment
+oc new-project ${USER}-stage-environment
+oc new-project ${USER}-cicd-environment
+```
+
+**_NOTE:_**  Continue this setup in `${USER}-cicd-envnironment` project.
```
Comment (bigkevmcd) [inline Example.md]: s/envnironment/environment/

File: Example.md [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+
+
+## Quay Credentials to push built image to Quay.io registry
+
+ Some of the Tasks in this Tutorial involve pulling and pushing images to Quay image registry.   (The image is to be built by CI Pipeline.)   Before we can start creating Kuberetes resources, we need to obtain credentials for your Quay user account.
```
Comment (bigkevmcd) [inline Example.md]: s/Kuberetes/Kubernetes/

File: Example.md [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+
+ * Select `Edit Repository Permissions` menu item
+
+  ![Screenshot](edit-token-permission.png)
+
+* Grant `write` permission to repository `taxi`
+
+ ![Screenshot](grant-write-permission.png)
+
+ * Download Kubernates Pull Secret (\<Quay.io user\>-robot-secret.yml)
```
Comment (bigkevmcd) [inline Example.md]: s/Kubernates/Kubernetes/ in a number of places.

File: Example.md [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+```shell
+oc apply -f <path/to>/${QUAY_USER}-robot-secret.yml
+  ```
+* Download Docker Configuration file to `<Quay user>-robot-auth.json`
+
+   ![Screenshot](quay-download-docker-config.png)
+
+* Create `regcred` secret from Docker Configuration file
+
+**_NOTE:_**  Make sure `QUAY_USER` envrionment variable is set.
```
Comment (bigkevmcd) [inline Example.md:71]: s/envrionment/environment/ there are a few cases of this.

File: Example.md [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+
+ * Download Kubernates Pull Secret (\<Quay.io user\>-robot-secret.yml)
+
+ ![Screenshot](quay-download-secret.png)
+
+ * Create Kubernates secret from downloaded pull secret yaml
+
+ **_NOTE:_**  Replace \<path\to\> with path to yaml file. 
+```shell
+oc apply -f <path/to>/${QUAY_USER}-robot-secret.yml
```
Comment (bigkevmcd) [inline Example.md]: I don't think you need this?

File: Example.md [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+
+We need to allow `demo-sa` to run as root in container.   This blog may help your understanding on Security Context Constraints. 
+[Understanding Service Accounts and SCCs](https://blog.openshift.com/understanding-service-accounts-sccs/).  We also need to bind`Edit`Role to `demo-sa`.  The second command below is another way to perform RoleBinding.  See [Managing Role-based Access Control (RBAC)](https://docs.openshift.com/container-platform/3.11/admin_guide/manage_rbac.html)
+
+Run the following commands setup additional security polcies for `demo-sa`.
+ ```shell
+oc adm policy add-scc-to-user privileged -z demo-sa
+oc adm policy add-role-to-user edit -z demo-sa
+```
+
```
Comment (bigkevmcd) [inline Example.md:166]: Could perhaps indicate _why_ this permission is needed?

File: Example.md [hunk 9]
```diff
@@ -0,0 +1,977 @@
...
+oc create rolebinding demo-sa-admin-dev --clusterrole=admin --serviceaccount=${USER}-cicd-environment:demo-sa --namespace=${USER}-dev-environment
+oc create rolebinding demo-sa-admin-stage --clusterrole=admin --serviceaccount=${USER}-cicd-environment:demo-sa --namespace=${USER}-stage-environment
+```
+## Create Tasks
+
+### buildah-task
+
+The following is a replica of [buildah Task](https://github.com/tektoncd/catalog/tree/master/buildah).  You can read the documentation there for the usage of the Task.  In a nutshell, it has a `build` step and a `push` step.  It builds an image from source and push the built image to a registry.
+
+Create buildah task yaml file with the following content.  Apply yaml file.   `oc appyl -f <yaml file>`
```
Comment (bigkevmcd) [inline Example.md]: s/applyl/apply/ and a couple of cases of this.

File: Example.md [hunk 10]
```diff
@@ -0,0 +1,977 @@
...
+      - "--target-url"
+      - "$(inputs.params.TARGET_URL)"
+      - "--description"
+      - "$(inputs.params.DESCRIPTION)"
+      - "--context"
+      - "$(inputs.params.CONTEXT)"
+```      
+* Run the following command to apply `temp.yaml` with `sed` command to replace string value
+
+**_NOTE:_**  Make sure `GIBHUB_USER` envrionment variable is set.
```
Comment (bigkevmcd) [inline Example.md:302]: s/GIBHUB/GITHUB/

File: Example.md [hunk 11]
```diff
@@ -0,0 +1,977 @@
...
+
+```shell
+cat temp.yaml | sed s"/\$GITHUB_USER/$GITHUB_USER/" | oc apply -f -
+```
+
+### deploy-from-source-task
+
+This command runs `kubectl -k` to deploy manifacts with [kuctomize](https://github.com/kubernetes-sigs/kustomize) option.  It executes commands on image quay.io/kmcdermo/k8s-kubectl.   The source is located in https://github.com/bigkevmcd/k8s-kubectl.
```
Comment (bigkevmcd) [inline Example.md:310]: s/manifacts/manifests/

s/kuctomize/kustomize/

Couple of cases of this.

File: Example.md [hunk 12]
```diff
@@ -0,0 +1,977 @@
...
+        - "--dry-run=$(inputs.params.DRYRUN)"        
+        - "-n"
+        - "$(inputs.params.NAMESPACE)"
+        - "-k"
+        - "$(inputs.params.PATHTODEPLOYMENT)"
+```
+
+## Create Trigger Templates and Trigger Bindings
+
+The relationship between Trigger Binding, Trigger Template and EventListener are as follow.   EventListener is a K8s Service that listens for external events (such as WebHook Event).  EventListener is configurated with actions to perform when matched events are received.  An action can trigger a PipelineRun.  In EventListener, it matches events and routes them to Trigger Binding/Template pairs.  Event message, Trigger Binding, and Trigger Template together are all we need to trigger a PipelineRun.  In order to start a PipelineRun, you will need PipelineResources to feed the PipelineRun.  Trigger Binding knows how to pick Event message's fields and use Trigger Template to create PipelineResources.  The target Pipeline information is in the Trigger Template.  This is how Trigger Binding, Tirgger Template, and EventListener work together to start a PipelineRun upon receiving an Event.  For more information, please see https://github.com/tektoncd/triggers
```
Comment (bigkevmcd) [inline Example.md:417]: I'd maybe say that an EventListener receives incoming HTTP POST requests.

File: Example.md [hunk 13]
```diff
@@ -0,0 +1,977 @@
...
+kind: EventListener
+metadata:
+  name: cicd-event-listener
+spec:
+  serviceAccountName: demo-sa
+  triggers:
+    - name: dev-ci-build-from-pr
+      interceptors:
+        - cel:
+            filter: (header.match("X-GitHub-Event", "pull_request") && body.action == "opened" || body.action == "synchronize") && body.pull_request.head.repo.full_name == "$GITHUB_USER/taxi"
```
Comment (bigkevmcd) [inline Example.md:908]: I've updated the event-listener YAML file, you'll want to replace `"` with `'`.

File: Example.md [hunk 14]
```diff
@@ -0,0 +1,977 @@
...
+
+```
+
+* Run the following command to replace strings and apply yaml
+
+**_NOTE:_**  Make sure `GITHUB_USER` envrionment variable is set.
+
+```shell
+cat temp.yaml  | sed s"/\$GITHUB_USER/$GITHUB_USER/" | oc apply -f -
+```
```
Comment (bigkevmcd) [inline Example.md:946]: Something that isn't clear, is why sometimes use `${}` templating, and other times use `sed` ?

File: Example.md [hunk 15]
```diff
@@ -0,0 +1,977 @@
...
+# Test it
+
+## Generate a Webhook Event
+
+* Submit a Pull Request to your `taxi` GitHub repository
+* Merge the Pull Request which should trigger a Pipeline Run in the CI Pipeline
+
+## Re-generate Webhook Event without Pull Request
+
+* Go back to the WebHooks Setting in your `taxi` GitHub repository and look for the last `Push` event.  You can re-deliever the event for testing without creating a Pull Request.
```
Comment (bigkevmcd) [inline Example.md]: s/deliever/deliver/

File: Example.md [hunk 16]
```diff
@@ -0,0 +1,977 @@
...
+  wildcardPolicy: None
+EOF
+```
+## Create github-auth secret
+
+Create secret with the github token that you have regenerated/downloaded.
+
+**_NOTE:_**  Make sure `GITHUB_USER` envrionment variable is set.
+```shell
+oc create secret generic ${GITHUB_USER}-github-auth --from-file="<path/to>/github-token.txt"
```
Comment (bigkevmcd) [inline Example.md]: You're creating this in a named environment, do you need a named github-auth token too?

File: Example.md [hunk 17]
```diff
@@ -0,0 +1,977 @@
...
+```
+
+## Create a Webhook
+
+Find out the `Payload URL` to be used in WebHook.   Run the following command to identify the external URL of the `github-webhook-event-listener`.   Add `http://` to the host to form `Payload URL` for Webhook.
+
```
Comment (bigkevmcd) [inline Example.md:991]: Probably should mention setting it to `JSON` rather than form encoding?

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: Example.md (hunk 1 above)
Comment (wtam2018) [reply Example.md]: Github web UI offers import from existing repos as one of the options.   I'll add text to mention following Github option to import repos.

File: Example.md (hunk 17 above)
Comment (wtam2018) [reply Example.md:991]: It is highlighted in the screen shot but I can explicitly call it out.

File: Example.md (hunk 16 above)
Comment (wtam2018) [reply Example.md]: You don't have to have it.  I can remove it.

File: Example.md (hunk 13 above)
Comment (wtam2018) [reply Example.md:908]: If this works (which it does), then I would rather leave it.

File: Example.md (hunk 14 above)
Comment (wtam2018) [reply Example.md:946]: I got rid of ${}

**Pull request merged**